File: src/main/java/com/smartral/fcm_notifier/Message.java
package com.smartral.fcm_notifier;

import com.google.gson.annotations.SerializedName;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by @author ali on 8/9/19
 */
public class Message {

    @SerializedName("registration_ids")
    private List<String> mTokens;

    @SerializedName("data")
    private Map<String, Object> mData;

    @SerializedName("priority")
    private String mPriority;

    @SerializedName("time_to_live")
    private Integer mTimeToLive = 60;

    private Message(Builder mBuilder) {
        this.mTokens = mBuilder.mTokens;
        this.mData = mBuilder.mData;
        this.mPriority = mBuilder.mPriority.mValue;
        this.mTimeToLive = mBuilder.mTimeToLive;
    }

    public enum MessagePriority {
        HIGH("HIGH"),
        NORMAL("NORMAL");

        private String mValue;

        MessagePriority(String mValue) {
            this.mValue = mValue;
        }

        public String getmValue() {
            return mValue;
        }
    }

    public static class Builder {

        List<String> mTokens;
        Map<String, Object> mData = new HashMap<>();
        MessagePriority mPriority;
        Integer mTimeToLive;

        public Builder setTokens(List<String> mTokens) {
            this.mTokens = mTokens;
            return this;
        }

        public Builder addData(String key, Object data) {
            this.mData.put(key, data);
            return this;
        }

        public Builder setPriority(MessagePriority mPriority) {
            this.mPriority = mPriority;
            return this;
        }

        public Builder setTimeToLive(Integer mTimeToLive) {
            this.mTimeToLive = mTimeToLive;
            return this;
        }

        public Message build() {
            return new Message(this);
        }
    }
}

File: src/main/kotlin/Message.kt
import com.google.gson.annotations.SerializedName

/**
 * Created by @author ali on 8/9/19
 */
class Message(mBuilder: Builder) {

    @SerializedName("registration_ids")
    private var mTokens: List<String>? = null

    @SerializedName("data")
    private var mData: Map<String, Any>? = null

    @SerializedName("priority")
    private var mPriority: String? = null

    @SerializedName("time_to_live")
    private var mTimeToLive: Long = 60L

    init {
        this.mTokens = mBuilder.mTokens
        this.mData = mBuilder.mData
        this.mPriority = mBuilder.mPriority?.mValue
        this.mTimeToLive = mBuilder.mTimeToLive
    }

    class Builder {

        internal var mTokens: List<String>? = null
        internal var mData: Map<String, Any>? = null
        internal var mPriority: MessagePriority? = null
        internal var mTimeToLive: Long = 60L

        fun setTokens(mTokens: List<String>) : Message.Builder {
            this.mTokens = mTokens
            return this
        }

        fun setData(mData: Map<String, Any>?) : Message.Builder {
            this.mData = mData
            return this
        }

        fun setTimeToLive(mTimeToLive: Long) : Message.Builder {
            this.mTimeToLive = mTimeToLive
            return this
        }

        fun setPriority(mPriority: MessagePriority?) : Message.Builder {
            this.mPriority = mPriority
            return this
        }

        fun build() : Message {
            return Message(this)
        }
    }

    enum class MessagePriority(val mValue: String) {
        HIGH("HIGH"),
        NORMAL("NORMAL")
    }
}